pkg/modprovider/state.go

```diff
@@ -200,46 +227,105 @@ func (h *moduleStateHandler) Diff(
 		KeepOutputValues: true,
 	}
 
-	oldInputs, err := plugin.UnmarshalProperties(oldProps, opts)
+	newInputs, err := plugin.UnmarshalProperties(req.News, opts)
 	if err != nil {
 		return nil, err
 	}
 
-	newInputs, err := plugin.UnmarshalProperties(newProps, opts)
+	contract.Assertf(newInputs["moduleInputs"].IsObject(), "expected moduleInputs as an Object")
+	moduleInputs := newInputs["moduleInputs"].ObjectValue()
+
+	plan, err := mod.plan(ctx, logger, tf, moduleInputs, oldState)
 	if err != nil {
 		return nil, err
 	}
-	moduleInputDiff := !oldInputs["moduleInputs"].DeepEquals(newInputs["moduleInputs"])
 
-	if !newState.Equal(oldState) || moduleInputDiff {
-		changes = pulumirpc.DiffResponse_DIFF_SOME
+	if !plan.HasChanges() {
+		// Since Create/Update will not be called after this, to avoid hangs we need to make sure the unchanged
+		// TF state is propagated to the rest of the system. Currently this is done approximately by running a
+		// no-op apply for its side-effects.  It would be more efficient to do mod.planStore.SetState if
+		// we had the capacity to hydrate the State from the serialized form in oldState.
+		mod.apply(ctx, logger, tf)
```

Comment:
This is really annoying and a source of slowdown. Wonder if TF states or the result of terraform state pull can be parsed into a tfjson.State.